pkg/apis/sessions/session_state.go

```diff
@@ -30,9 +33,30 @@ type SessionState struct {
 	PreferredUsername string   `msgpack:"pu,omitempty"`
 }
 
+// CreatedAtNow sets a SessionState's CreatedAt to now
```

Comment:
Open to suggestions on this.

This was all over the place in the providers with embedded direct `time.Now()` calls making session timing logic within them impossible to mock effectively.

I want to try to take more control of setting the session pieces here in a uniform way instead of stuff just dumping fields into the public struct. Thoughts?

Reply:
So sessions is where I wanted to talk about our time mocking posture more in general. These are consumed throughout a bunch of our other packages. For them to more accurately test session expiration scenarios, we likely need an entry point to fast forward session time (without access to the underlying session because it might be internal to the other package we might be consuming - e.g. session storage, cookies).

Reply:
Here's the brain dump I put into Slack a bit ago:

I want to talk more about redesigning how we use time.Now in the project and how we expose it in a convenient way that allows for good tests (we’re probably lacking on border expiration unit test cases because we can’t mock time easily in some areas).

Options: now time.Time method parameters (likely adds code climate violations), nowFunc func() time.Time fields on structs (a little bloat for widely used objects, only available within package tests due to private), var nowFunc = time.Time private package variables (again only available within package tests), or var NowFunc = time.Time public package variables (a little sloppy exposing this, but gives the best coverage for testing — particularly integration tests we are missing from the project where we want to fully mock the flow of time).

Want me to throw together an RFC issue so we can discuss this more publicly? After integrating mockoidc with jwt-go I am starting to lean heavily toward the last one — because they exposed time.Now publically to override, I could completely mock time and test border cases for JWT expiration & refresh scenarios in end to end integration tests.

Reply:
> My thinking here is that you ensure that the NowFunc only affects the current session you're manipulating in the tests, other tests could be run in parallel etc without worry

Ah parallel tests... didn't even think about that. Yes, that makes package level time functions a much less desirable way to manage this 🤔

Reply:
Handled in Clock package now, resolving.